### ISSUE TITLE
WCAG20 compatibility

### ISSUE THREAD

**Issue by Noino** [surveyjs/surveyjs#601]
### Are you requesting a feature, reporting a bug or ask a question?
Bug/Feature

### What is the current behavior?
The html generated does not conform to WCAG20 (https://www.w3.org/TR/WCAG20/) standards, and it therefor unusable by people with certain disabilities

### What is the expected behavior?
That the html generated would confirm to WCAG20 and be usable by everyone.

### How would you reproduce the current behavior (if this is a bug)?
Make a form, any kind of form, copy the resulting html to a validator
here: https://achecker.ca/checker/index.php
or using an extension: https://chrome.google.com/webstore/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh

#### Provide the test code and the tested page URL (if applicable)

Tested page URL:

Test code

```
var surveyJSON = {pages:[{name:"page1",elements:[{type:"paneldynamic",name:"question12",templateElements:[{type:"file",name:"question7"},{type:"matrix",name:"question8",columns:["Column 1","Column 2","Column 3"],rows:["Row 1","Row 2"]}],panelCount:1},{type:"matrixdropdown",name:"question9",columns:[{name:"Column 1"},{name:"Column 2"},{name:"Column 3"}],choices:[1,2,3,4,5],rows:["Row 1","Row 2"]},{type:"matrixdynamic",name:"question10",columns:[{name:"Column 1"},{name:"Column 2"},{name:"Column 3"}],choices:[1,2,3,4,5]},{type:"panel",name:"panel1",elements:[{type:"multipletext",name:"question11",items:[{name:"text1"},{name:"text2"}]}]},{type:"text",name:"question1"},{type:"checkbox",name:"question2",choices:["item1","item2","item3"]},{type:"radiogroup",name:"question3",choices:["item1","item2","item3"]},{type:"dropdown",name:"question4",choices:["item1","item2","item3"]},{type:"comment",name:"question5"},{type:"rating",name:"question6"}]}]}

function sendDataToServer(survey) {
  var resultAsString = JSON.stringify(survey.data);
  alert(resultAsString); //send Ajax request to your web server.
}

var survey = new Survey.Model(surveyJSON);
$("#surveyContainer").Survey({
    model: survey,
    onComplete: sendDataToServer
});


```

### Specify your

* browser:
* browser version:
* surveyjs platform (angular or react or jquery or knockout or vue): jquery
* surveyjs version: 0.12.23

**Comment by andrewtelnov** [comment]
@Noino We will work on this. I think it is important to support WCAG20. We will discuss this issue and I will comeback.

Thank you,
Andrew

**Comment by Noino** [comment]
@andrewtelnov Woah, quick answer too.
Sounds good. The problems that my example form has according to the validator i linked is mostly just the way labels are used, WCAG20 requires labels to use the 'for' attribute to refer to their linked input, so don't think the fix would be all too much of a rewrite.

**Comment by tsv2013** [comment]
@Noino I've fixed a set of compatibility issues. Could you check it in the nearest release?

**Comment by Noino** [comment]
@tsv2013 Sorry for the delayed response.
I updated to 0.12.32 and ran the same form through the validators i linked. It seems there are still some problems.

The chrome addon reports this code block
```
<label class="" style="margin: 0px; position: absolute;">
<input id="sq_113i" type="radio" class="" name="question8_Row 1" value="Column 1" aria-label="question8">
<span class="circle">
<span class="check">
</label>
```
Generated from this json
```
{type:"matrix",name:"question8",columns:["Column 1","Column 2","Column 3"],rows:["Row 1","Row 2"]}
```

as flagged with error:
> Empty form label
> **What It Means**
> A form label is present, but does not contain any content.
> **Why It Matters**
> A <label> element that is associated to a form control but does not contain text will not present any information about the form control to the user.
> **How to Fix It**
> Ensure that the form label contains text that describes the function of the associated form control. Labels are not required for image, submit, reset, button, or hidden form controls. If a label is not necessary visually, a descriptive title attribute may be added to the form control. 
> **Standards and Guidelines**
> Section 508 (n)
> 1.1.1 Non-text Content (Level A)
> 1.3.1 Info and Relationships (Level A)
> 2.4.6 Headings and Labels (Level AA)
> 3.3.2 Labels or Instructions (Level A)

And also alert:
> Missing fieldset
> **What It Means**
> A group of check boxes or radio buttons is not enclosed in a fieldset.
> **Why It Matters**
> A fieldset provides a visual and structural grouping of related form elements. It is typically necessary for groups of check boxes or radio buttons where a higher level description (called a legend) is necessary to understand the function of the check boxes or radio buttons. The description will be identified by a screen reader only if provided in a fieldset legend.
> **Standards and Guidelines**
> Section 508 (n)
> 1.1.1 Non-text Content (Level A)
> 1.3.1 Info and Relationships (Level A)
> 2.4.6 Headings and Labels (Level AA)
> 3.3.2 Labels or Instructions (Level A)

There are some others listed too but these are the only errors it reports.


I've also attached the pdf report from achecker.ca which lists 60 known problems with this sample form.


[achecker_2017-10-10_10-22-45.pdf](https://github.com/surveyjs/surveyjs/files/1371496/achecker_2017-10-10_10-22-45.pdf)

**Comment by tsv2013** [comment]
@Noino This markup
```
<label class="" style="margin: 0px; position: absolute;">
<input id="sq_113i" type="radio" class="" name="question8_Row 1" value="Column 1" aria-label="question8">
<span class="circle">
<span class="check">
</label>
```
Is used to support bootstrap material design skin, because it requires specific DOM structure. We need to discuss it with the team.

**Comment by tsv2013** [comment]
@Noino @andrewtelnov I've added dummy span as a temporary measure to suppress the error. As for fieldset tag - if this warning is not crucial, I wouldn't like to add extra markup on the page.

**Comment by Noino** [comment]
I believe that it is necessary to fieldset radio buttons, and even checkboxes that are part of a multiple choice question.

> Grouping controls is most important for related radio buttons and checkboxes. A set of radio buttons or checkboxes is related when they all submit values for a single named field. They work in the same way as selection lists, allowing the user to choose from a set of options, except selection lists are single controls while radio buttons and checkboxes are multiple controls. The individual label associated with each radio or checkbox control may not fully convey the group's descriptive context. In this situation, it is essential that they be grouped together semantically to facilitate being treated as a single control, as well as to provide an additional group level description.
https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/H71

**Comment by okkuweb** [comment]
What is the status of this issue? It seems like there has been some progress, but the changes are still underway?

**Comment by tsv2013** [comment]
@okkuweb I've fixed all errors shown by WCAG20 validators. There are still some warnings. Some of them require markup changes and adding additional markup. The markup changes can became "breaking changes". I'd like to evaluate warnings cruciality to fix most crucial ones to minimize markup changes.

**Comment by Noino** [comment]
@tsv2013 Thanks for your hard work.

Seems you've got pretty much all of them down. Really only fieldsets remaining as far as i can see, and i'm not sure how you are supposed to accomplish that within matrix tables, but in my form "6. question2" and "7. question3" you have a <form> tag where <fieldset> would seem natural.

One thing i noticed trying that same form through achecker.ca again (even though it seems to not understand html5 very well) was that I'm getting a duplicate id "sp_100" which i don't think is valid html.
root level div
<div id="sp_100" class="sv_body">
and "4. question11" first text box
<input id="sp_100" class="sv_q_mt_item_value" type="text" placeholder="" aria-label="text1">

**Comment by tsv2013** [comment]
@Noino Thank you for the valueable feedback! I'll fix remaining issues and update this thread. Hope it will be soon.

**Comment by tsv2013** [comment]
@Noino I've added fieldset tags with necessary 'legend' tags inside.
I had to implement not so beautiful solution for matrixes as I wanted to. I wrapped with fieldset tag the whole matrix question due to the 'tr' and 'td' tags wrapping crashes table structure.
Changes wil be available in the nearest update.

Thank you for cooperation

**Comment by Noino** [comment]
The lastest change (8578f4f) seems to be missing in jquery and angular?

**Comment by tsv2013** [comment]
@Noino  jquery and angular libraries are based on the the react code. They're just wrappers.

**Comment by Noino** [comment]
@tsv2013 Alright. Thought it might have been the case since i was seeing some complaints about fieldsets missing still, but turns out it was a cache issue, hard reload fixed almost all.
Still, one type that i clearly missed in my previous response, "10. question6" i get error for due to missing fieldset. This is the rating type question.

Sorry for resurrecting a closed ticket, should've noticed this before
